src/Filters/PermissionFilter.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace CodeIgniter\Shield\Filters;
+
+use CodeIgniter\Filters\FilterInterface;
+use CodeIgniter\HTTP\RequestInterface;
+use CodeIgniter\HTTP\Response;
+use CodeIgniter\HTTP\ResponseInterface;
+
+use CodeIgniter\Shield\Exceptions\PermissionException;
+
+class PermissionFilter implements FilterInterface
+{
+    /**
+     * Do whatever processing this filter needs to do.
+     * By default it should not return anything during
+     * normal execution. However, when an abnormal state
+     * is found, it should return an instance of
+     * CodeIgniter\HTTP\Response. If it does, script
+     * execution will end and that Response will be
+     * sent back to the client, allowing for error pages,
+     * redirects, etc.
+     *
+     * @param array|null $arguments
+     *
+     * @return mixed
+     */
+    public function before(RequestInterface $request, $arguments = null)
+    {
+        if (empty($arguments)) {
+            return;
+        }
+
+        if (! function_exists('auth')) {
+            helper('auth');
+        }
+
+        if (! auth()->loggedIn()) {
+            return redirect()->to('login');
+        }
+
+        $result = true;
+
+        foreach ($arguments as $permission) {
+            $result = $result && auth()->user()->can($permission);
+        }
+
+        if (! $result) {
+            throw PermissionException::forUnknownPermission(...$arguments);
+        }
```

Comment:
If any single check fails you can short-circuit. My suggestion:
```suggestion
        foreach ($arguments as $permission) {
            if (! auth()->user()->can($permission)) {
                throw PermissionException::forWhatever($permission);
            }
        }
```
(Though the exception type will need to change.)